로그인 기능

mingang211 opened the pull request:
## #️⃣ Issue Number

#2 로그인 기능

## 📝 요약(Summary)
**feat**
- 로그인을 하기 위한 LoginForm 작성
- Controller 와 Service 작성
- UserRepository에서 email을 통해 패스워드가 동일한지, 이메일 인증이 되었는지 확인
- ErrorCode에     LOGIN_CHECK_FAIL(HttpStatus.BAD_REQUEST, "아이디나 패스워드를 확인해 주세요") 추가
- Aes256Util 생성하여 암호화 복호화 만들기
- 로그인시 USER, ADMIN으로 분류
- email과 password 확인 후 token 생성
- 유저정보와 토큰 인증 만료 확인 기능 작성
- JwtConfig class를 생성하여 JwtAuthenticationProvider Bean등록
- token을 이용하여 유저정보를 불러와서 db에서 유저정보를 가져오고 없을 시 NOT_FOUND_USER 예외 발생
- UserDto로 정보 교환
- Filter를 사용해 '/user/*'경론는 로그인시만 접근할 수 있도록 하였고 token이 없을시, 토큰 기한이 지났을 시, 토큰의 유저 email과 id가 다를시 예외 발생


**test:**
- JwtAuthenticationProvider, UserControllerTest,  UserLogInController에 대한 테스트 작성
- 실패 test는 시간이 너무 오래 걸릴 것 같아 보류
-  Aes256Util 암호화 복호화 테스트


**refact:**
- 암호화 key, iv, jwt secretkey properties에 저장
- Loginform에 어노테이션 추가
-  Aes256Util 수정 -> 기존 코드로 로그인시 Key와 vi가 제대로 받아지지 않아 JWTConfig를 만들어 key를 받아오게 수정

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 📸스크린샷 (선택)

## 💬 공유사항 to 리뷰어
- Test 메서드 명을 작성할 때 어떻게 작성하는 게 좋은지 잘 모르겠습니다. 혹시 추천하시거나 참고할 만한 자료있으시다면 공유 부탁드립니다.
- Spring-security를 사용해 로그인을 구현을 해보고 싶었는데 이해하기가 조금 어려워 현재 코드처럼 작성했습니다.

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

mingang211 committed:
#2 feat:002-User-Login 로그인 기능 -> 이메일을 통해 비밀번호와 이메일 인증 확인을 한다.
- 로그인을 하기 위한 LoginForm 작성
- Controller 와 Service 작성
- UserRepository에서 email을 통해 패스워드가 동일한지, 이메일 인증이 되었는지 확인
- ErrorCode에     LOGIN_CHECK_FAIL(HttpStatus.BAD_REQUEST, "아이디나 패스워드를 확인해 주세요") 추가

mingang211 committed:
#2 feat:002-User-Login 로그인 기능 -> 암호화 복호화 확인
- Aes256Util 생성하여 암호화 복호화 만들기
- Aes256Util 암호화 복호화 테스트

mingang211 committed:
#2 feat:002-User-Login 로그인 기능 -> Jwt token 생성
- 로그인시 USER, ADMIN으로 분류
- email과 password 확인 후 token 생성
- 유저정보와 토큰 인증 만료 확인 기능 작성
- JwtConfig class를 생성하여 JwtAuthenticationProvider Bean등록
refact : 암호화 key, iv, jwt secretkey properties에 저장

mingang211 committed:
#2 feat:002-User-Login 로그인 기능 -> Jwt으로 로그인 하여 회원 정보 불러오기
- token을 이용하여 유저정보를 불러와서 db에서 유저정보를 가져오고 없을 시 NOT_FOUND_USER 예외 발생
- UserDto로 정보 교환
- Filter를 사용해 '/user/*'경론는 로그인시만 접근할 수 있도록 하였고 token이 없을시, 토큰 기한이 지났을 시, 토큰의 유저 email과 id가 다를시 예외 발생
refact : Aes256Util 수정 -> 기존 코드로 로그인시 Key와 vi가 제대로 받아지지 않아 JWTConfig를 만들어 key를 받아오게 수정

mingang211 committed:
#2 test:002-User-Login 로그인 기능 테스트-> JwtAuthenticationProvider, UserControllerTest, UserLogInController에 대한 테스트 작성
- 실패 test는 시간이 너무 오래 걸릴 것 같아 보류
refact : Loginform에 어노테이션 추가

seulahhh commented on src/main/java/com/zerobase/zbfintech/util/Aes256Util.java:
```diff
@@ -0,0 +1,54 @@
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+
+@Component
```
static 메서드는 component(spring bean) 등록 없이 바로 클래스 이름을 통해 직접 호출 가능한 것으로 알고 있는데, 혹시 빈으로도 등록하시고 static으로 메서드를 사용하시는 이유를 알 수 있을까요!!

spring-kotlin-lover commented on src/main/java/com/zerobase/zbfintech/controller/UserController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Operation(summary = "회원 정보 가져오기", description = "로그인한 회원의 정보를 가져옵니다.")
+    @GetMapping("/getInfo")
+    public ResponseEntity<UserDto> getInfo(@RequestHeader(name="Authorization") String token){
```
Authorization Header 가져올때 Spring Core 에 있는 `HttpHeaders` 클래스 활용하시면 좋겠네요~!

spring-kotlin-lover commented on src/main/java/com/zerobase/zbfintech/controller/UserLogInController.java:
```diff
@@ -0,0 +1,26 @@
+    }
+    @Operation(summary = "로그인", description = "로그인 요청을 보내고 tokem을 생성합니다")
+    @PostMapping("/user")
+    public ResponseEntity<String> userLogIn(@RequestBody LoginForm loginForm) {
```
습관화가 아직 덜 된것 같아서, 다시 코멘트 드립니다. spring boot validation 을 이용하여 사용자 request body 에 대한 유효성 검사는 꼭 실시해주세요

spring-kotlin-lover commented on src/main/java/com/zerobase/zbfintech/filter/UserFilter.java:
```diff
@@ -0,0 +1,40 @@
+            throw new ServletException("Authorization token is missing");
+        }
+
+        if(!jwtAuthenticationProvider.validateToken(token)){
+            throw new ServletException("Invalid Access");
+        }
+        UserVo vo = jwtAuthenticationProvider.getUserVo(token);
```
Provider 의 메서드 호출순서를 꼭 지켜야 올바르게 UserVo 를 파싱할 것으로 보이네요. 제 생각에는 현재 로직상 불필요한 jwt parse 가 2번 일어나게 되니, getValidatedUserVo 라는 메서드명으로 정의하고 2개의 기능을 합쳐보는 것도 좋을 것 같습니다.

리턴값은 Optional 로 감싸서 Optional 이 비어있다면, 예외를 던지는 방향으로 해도 괜찮을 것 같네요.

spring-kotlin-lover commented on src/main/java/com/zerobase/zbfintech/service/UserLogInService.java:
```diff
@@ -0,0 +1,39 @@
+        return provider.createToken(loginUser.getEmail(),loginUser.getId(), UserType.USER);
+    }
+
+    public Optional<User> findValidUser(String email, String password) {
+        return userRepository.findByEmail(email).stream()
+                .filter(user -> user.getPassword().equals(password)
+                && user.is_email_verified())
```
java 의 기본 네이밍 규칙은 camelCase 입니다. snake_case 가 되지 않도록 개선해주세요

mingang211 replied:
감사합니다 해당 부분 camelCase로 수정했습니다.

spring-kotlin-lover commented on src/main/java/com/zerobase/zbfintech/util/Aes256Util.java:
```diff
@@ -0,0 +1,54 @@
+            return Base64.encodeBase64String(encrypted);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
```
예외를 catch 하여 return null 으로 작성하는 습관은 안좋은 로직 전개방식중의 하나입니다. 호출처에서 매번 null 체크를 해야되고, null 로 리턴하게 된 계기가 정말 데이터가 없어서 null 인지 분간이 어렵기 때문인데요.

가급적이면 예외가 발생한다면 예외를 전파하는게 좋을 것 같습니다. 예외를 핸들링하는 쪽에서 어떻게 처리할 것인지 전개해나가는 습관을 들여보면 좋을 것 같아요.
https://ducktopia.tistory.com/111

spring-kotlin-lover commented on src/test/java/com/zerobase/zbfintech/auth/JwtAuthenticationProviderTest.java:
```diff
@@ -0,0 +1,92 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+
+@SpringBootTest
+public class JwtAuthenticationProviderTest {
```
단위 테스트와 통합 테스트의 차이 부터 이해해보시면 좋을 것 같습니다. 언제는 단위테스트를 사용하고 언제는 통합테스트를 사용할까요? 2개의 테스트가 갖는 장단점에 한번 생각해보실만한 자료를 드릴테니, 공부하여 개념 정립해보면 좋을 것 같네요.

https://curiousjinan.tistory.com/entry/integration-vs-unit-testing

mingang211 replied:
@SpringBootTest 어노테이션을 남발한다면 앱이 무거워진다는 점을 배웠습니다. 현재 코멘트 주신 부분은 @Value로 값을 가져오기에 @SpringBootTest를 사용해야 할 것 같습니다.

mingang211 committed:
#2 refactor:002-User-Login-> 코드 리뷰 반영
-AES_ENCRYPTION_ERROR,AES_DECRYPTION_ERROR 추가
- provider에 getValidatedUserVo명령어 추가하여 토큰 유효성검증과 유저 정보를 불러옴
_ LoginForm에 Vaildation추가
- is_email_verified -> isEmailVerified snake case에서 camel case로 변경
_ HttpHeaders에서 token을 가져옴

mingang211 committed:
#2 refactor:002-User-Login-> 코드 리뷰 반영
-AES_ENCRYPTION_ERROR,AES_DECRYPTION_ERROR 추가
- provider에 getValidatedUserVo명령어 추가하여 토큰 유효성검증과 유저 정보를 불러옴
_ LoginForm에 Vaildation추가
- is_email_verified -> isEmailVerified snake case에서 camel case로 변경
_ HttpHeaders에서 token을 가져옴

mingang211 committed:
Merge remote-tracking branch 'origin/002-User-Login' into 002-User-Login
